compiler/rustc_errors/src/lib.rs

```diff
@@ -1091,8 +1092,9 @@ impl DiagCtxt {
 
     /// Used when trimmed_def_paths is called and we must produce a diagnostic
     /// to justify its cost.
+    #[track_caller]
     pub fn set_must_produce_diag(&self) {
-        self.inner.borrow_mut().must_produce_diag = true;
+        self.inner.borrow_mut().must_produce_diag = Some(Backtrace::capture());
```

Comment:
I think this should only be called once? Might be worth asserting that it is `None` first, i.e. that we only capture a backtrace once.